Properly pass variables into extracted function

We always pass variables by value and it can lead to incorrect code.
We should correctly determine how to pass variable: by value or by reference.

```rust
fn foo() {
    let vec = vec![1, 2, 3];
    <selection>println!("{}", vec.len());</selection>
    println!("{}", vec[0]);
}
```

**Expected:**
```rust
fn foo() {
    let vec = vec![1, 2, 3];
    bar(&vec);
    println!("{}", vec[0]);
}

fn bar(vec: &Vec<i32>) {
    println!("{}", vec.len());
}
```

**Actual:**
```rust
fn foo() {
    let vec = vec![1, 2, 3];
    bar(vec);
    println!("{}", vec[0]); // error[E0382]: use of moved value: `vec`
}

fn bar(vec: Vec<i32>) {
    println!("{}", vec.len()); 
}
```

--- a/src/main/kotlin/org/rust/lang/refactoring/extractFunction/RsExtractFunctionConfig.kt
+++ b/src/main/kotlin/org/rust/lang/refactoring/extractFunction/RsExtractFunctionConfig.kt
@@ -11,10 +11,7 @@ import com.intellij.psi.search.LocalSearchScope
 import com.intellij.psi.search.searches.ReferencesSearch
 import org.rust.ide.presentation.insertionSafeText
 import org.rust.ide.utils.findStatementsInRange
-import org.rust.lang.core.psi.RsExpr
-import org.rust.lang.core.psi.RsFunction
-import org.rust.lang.core.psi.RsPatBinding
-import org.rust.lang.core.psi.RsTypeParameter
+import org.rust.lang.core.psi.*
 import org.rust.lang.core.psi.ext.*
 import org.rust.lang.core.resolve.ImplLookup
 import org.rust.lang.core.types.ty.*
@@ -42,21 +39,23 @@ class ReturnValue(val expression: String?, val type: Ty) {
 class Parameter(
     val name: String,
     val type: Ty? = null,
-    val reference: Reference = Reference.NONE
+    val reference: Reference = Reference.NONE,
+    isMutableValue: Boolean = false
 ) {
 
     enum class Reference(val text: String) {
-        MUTABLE("&mut "), BORROWING("& "), NONE("")
+        MUTABLE("&mut "), IMMUTABLE("& "), NONE("")
     }
 
     companion object {
-        fun direct(value: RsPatBinding, requiredBorrowing: Boolean): Parameter {
+        fun direct(value: RsPatBinding, requiredBorrowing: Boolean, requiredMutableValue: Boolean): Parameter {
             val reference = when {
+                requiredMutableValue -> Reference.NONE
                 value.mutability.isMut -> Reference.MUTABLE
-                requiredBorrowing -> Reference.BORROWING
+                requiredBorrowing -> Reference.IMMUTABLE
                 else -> Reference.NONE
             }
-            return Parameter(value.referenceName, value.type, reference)
+            return Parameter(value.referenceName, value.type, reference, requiredMutableValue)
         }
 
         fun self(value: String): Parameter {
@@ -64,8 +63,10 @@ class Parameter(
         }
     }
 
+    private val mut = if (isMutableValue) "mut " else ""
+
     val parameterText: String
-        get() = if (type != null) "$name: ${reference.text}${type.insertionSafeText}" else name
+        get() = if (type != null) "$mut$name: ${reference.text}${type.insertionSafeText}" else name
 
     val argumentText: String
         get() = "${reference.text}$name"
@@ -162,11 +163,21 @@ class RsExtractFunctionConfig private constructor(
                     if (it.textOffset > start) return@mapNotNull null
                     val result = ReferencesSearch.search(it, LocalSearchScope(fn))
                     if (result.none { it.element.textOffset in start..end }) return@mapNotNull null
+
                     val requiredBorrowing = result.any { it.element.textOffset > end }
                         && it.type !== TyBool && it.type !== TyChar && it.type !is TyInteger && it.type !is TyFloat
                         && it.type !is TyReference
                         && !implLookup.isCopy(it.type)
-                    Parameter.direct(it, requiredBorrowing)
+
+                    val requiredMutableValue = it.mutability.isMut && result.any {
+                        val element = it.element
+                        if (element.textOffset !in start..end) return@any false
+                        if (element.ancestorStrict<RsValueArgumentList>() == null) return@any false
+                        val unaryExpr = element.ancestorStrict<RsUnaryExpr>()
+                        unaryExpr == null || unaryExpr.operatorType == UnaryOperator.REF_MUT
+                    }
+
+                    Parameter.direct(it, requiredBorrowing, requiredMutableValue)
                 }
                 .toMutableList()
 

--- a/src/main/kotlin/org/rust/lang/refactoring/extractFunction/RsExtractFunctionConfig.kt
+++ b/src/main/kotlin/org/rust/lang/refactoring/extractFunction/RsExtractFunctionConfig.kt
@@ -17,10 +17,7 @@ import org.rust.lang.core.psi.RsPatBinding
 import org.rust.lang.core.psi.RsTypeParameter
 import org.rust.lang.core.psi.ext.*
 import org.rust.lang.core.resolve.ImplLookup
-import org.rust.lang.core.types.ty.Mutability
-import org.rust.lang.core.types.ty.Ty
-import org.rust.lang.core.types.ty.TyPrimitive
-import org.rust.lang.core.types.ty.TyTuple
+import org.rust.lang.core.types.ty.*
 import org.rust.lang.core.types.type
 
 class ReturnValue(val expression: String?, val type: Ty) {
@@ -42,11 +39,24 @@ class ReturnValue(val expression: String?, val type: Ty) {
     }
 }
 
-class Parameter(val name: String, val type: Ty? = null, val mutability: Mutability = Mutability.IMMUTABLE,
-                requiredBorrowing: Boolean = false) {
+class Parameter(
+    val name: String,
+    val type: Ty? = null,
+    val reference: Reference = Reference.NONE
+) {
+
+    enum class Reference(val text: String) {
+        MUTABLE("&mut "), BORROWING("& "), NONE("")
+    }
+
     companion object {
         fun direct(value: RsPatBinding, requiredBorrowing: Boolean): Parameter {
-            return Parameter(value.referenceName, value.type, value.mutability, requiredBorrowing)
+            val reference = when {
+                value.mutability.isMut -> Reference.MUTABLE
+                requiredBorrowing -> Reference.BORROWING
+                else -> Reference.NONE
+            }
+            return Parameter(value.referenceName, value.type, reference)
         }
 
         fun self(value: String): Parameter {
@@ -54,19 +64,11 @@ class Parameter(val name: String, val type: Ty? = null, val mutability: Mutabili
         }
     }
 
-    private val mutabilityText = if (!requiredBorrowing && mutability.isMut) "mut " else ""
-
-    private val borrowingText = when {
-        requiredBorrowing -> if (mutability.isMut) "&mut " else "&"
-        else -> ""
-    }
-
     val parameterText: String
-        get() = if (type != null) "$mutabilityText$name: $borrowingText${type.insertionSafeText}" else name
+        get() = if (type != null) "$name: ${reference.text}${type.insertionSafeText}" else name
 
     val argumentText: String
-        get() = "$borrowingText$name"
-
+        get() = "${reference.text}$name"
 }
 
 class RsExtractFunctionConfig private constructor(
@@ -154,13 +156,16 @@ class RsExtractFunctionConfig private constructor(
             val letBindings = fn.descendantsOfType<RsPatBinding>()
                 .filter { it.textOffset <= end }
 
+            val implLookup = ImplLookup.relativeTo(fn)
             val parameters = letBindings
                 .mapNotNull {
                     if (it.textOffset > start) return@mapNotNull null
                     val result = ReferencesSearch.search(it, LocalSearchScope(fn))
                     if (result.none { it.element.textOffset in start..end }) return@mapNotNull null
                     val requiredBorrowing = result.any { it.element.textOffset > end }
-                        && it.type !is TyPrimitive && !ImplLookup.relativeTo(it).isCopy(it.type)
+                        && it.type !== TyBool && it.type !== TyChar && it.type !is TyInteger && it.type !is TyFloat
+                        && it.type !is TyReference
+                        && !implLookup.isCopy(it.type)
                     Parameter.direct(it, requiredBorrowing)
                 }
                 .toMutableList()

--- a/src/main/kotlin/org/rust/lang/refactoring/extractFunction/RsExtractFunctionConfig.kt
+++ b/src/main/kotlin/org/rust/lang/refactoring/extractFunction/RsExtractFunctionConfig.kt
@@ -50,7 +50,7 @@ class Parameter(
     companion object {
         fun direct(value: RsPatBinding, requiredBorrowing: Boolean, requiredMutableValue: Boolean): Parameter {
             val reference = when {
-                requiredMutableValue -> Reference.NONE
+                requiredMutableValue -> if (requiredBorrowing) Reference.MUTABLE else Reference.NONE
                 value.mutability.isMut -> Reference.MUTABLE
                 requiredBorrowing -> Reference.IMMUTABLE
                 else -> Reference.NONE
@@ -162,19 +162,24 @@ class RsExtractFunctionConfig private constructor(
                 .mapNotNull {
                     if (it.textOffset > start) return@mapNotNull null
                     val result = ReferencesSearch.search(it, LocalSearchScope(fn))
-                    if (result.none { it.element.textOffset in start..end }) return@mapNotNull null
 
-                    val requiredBorrowing = result.any { it.element.textOffset > end }
-                        && it.type !== TyBool && it.type !== TyChar && it.type !is TyInteger && it.type !is TyFloat
+                    val targets = result.filter { it.element.textOffset in start..end }
+                    if (targets.isEmpty()) return@mapNotNull null
+
+                    val hasRefOperator = targets.any {
+                        val operatorType = (it.element.ancestorStrict<RsUnaryExpr>())?.operatorType
+                        operatorType == UnaryOperator.REF || operatorType == UnaryOperator.REF_MUT
+                    }
+                    val requiredBorrowing = hasRefOperator || result.any { it.element.textOffset > end }
+                        // We manually check primitive types because of ImplLookup.isCopy does not support them
+                        && it.type != TyBool && it.type != TyChar && it.type !is TyInteger && it.type !is TyFloat
                         && it.type !is TyReference
                         && !implLookup.isCopy(it.type)
 
-                    val requiredMutableValue = it.mutability.isMut && result.any {
-                        val element = it.element
-                        if (element.textOffset !in start..end) return@any false
-                        if (element.ancestorStrict<RsValueArgumentList>() == null) return@any false
-                        val unaryExpr = element.ancestorStrict<RsUnaryExpr>()
-                        unaryExpr == null || unaryExpr.operatorType == UnaryOperator.REF_MUT
+                    val requiredMutableValue = it.mutability.isMut && targets.any {
+                        if (it.element.ancestorStrict<RsValueArgumentList>() == null) return@any false
+                        val operatorType = it.element.ancestorStrict<RsUnaryExpr>()?.operatorType
+                        operatorType == null || operatorType == UnaryOperator.REF_MUT
                     }
 
                     Parameter.direct(it, requiredBorrowing, requiredMutableValue)

--- a/src/main/kotlin/org/rust/lang/refactoring/extractFunction/RsExtractFunctionConfig.kt
+++ b/src/main/kotlin/org/rust/lang/refactoring/extractFunction/RsExtractFunctionConfig.kt
@@ -16,8 +16,10 @@ import org.rust.lang.core.psi.RsFunction
 import org.rust.lang.core.psi.RsPatBinding
 import org.rust.lang.core.psi.RsTypeParameter
 import org.rust.lang.core.psi.ext.*
+import org.rust.lang.core.resolve.ImplLookup
 import org.rust.lang.core.types.ty.Mutability
 import org.rust.lang.core.types.ty.Ty
+import org.rust.lang.core.types.ty.TyPrimitive
 import org.rust.lang.core.types.ty.TyTuple
 import org.rust.lang.core.types.type
 
@@ -40,10 +42,11 @@ class ReturnValue(val expression: String?, val type: Ty) {
     }
 }
 
-class Parameter(val name: String, val type: Ty? = null, val mutability: Mutability = Mutability.IMMUTABLE) {
+class Parameter(val name: String, val type: Ty? = null, val mutability: Mutability = Mutability.IMMUTABLE,
+                requiredBorrowing: Boolean = false) {
     companion object {
-        fun direct(value: RsPatBinding): Parameter {
-            return Parameter(value.referenceName, value.type, value.mutability)
+        fun direct(value: RsPatBinding, requiredBorrowing: Boolean): Parameter {
+            return Parameter(value.referenceName, value.type, value.mutability, requiredBorrowing)
         }
 
         fun self(value: String): Parameter {
@@ -51,17 +54,19 @@ class Parameter(val name: String, val type: Ty? = null, val mutability: Mutabili
         }
     }
 
-    val text: String
-        get() = buildString {
-            if (type != null) {
-                if (mutability.isMut) {
-                    append("mut ")
-                }
-                append("$name: ${type.insertionSafeText}")
-            } else {
-                append(name)
-            }
-        }
+    private val mutabilityText = if (!requiredBorrowing && mutability.isMut) "mut " else ""
+
+    private val borrowingText = when {
+        requiredBorrowing -> if (mutability.isMut) "&mut " else "&"
+        else -> ""
+    }
+
+    val parameterText: String
+        get() = if (type != null) "$mutabilityText$name: $borrowingText${type.insertionSafeText}" else name
+
+    val argumentText: String
+        get() = "$borrowingText$name"
+
 }
 
 class RsExtractFunctionConfig private constructor(
@@ -74,10 +79,10 @@ class RsExtractFunctionConfig private constructor(
 ) {
 
     private val parametersText: String
-        get() = parameters.joinToString(",") { it.text }
+        get() = parameters.joinToString(",") { it.parameterText }
 
     val argumentsText: String
-        get() = parameters.filter { it.type != null }.joinToString(",") { it.name }
+        get() = parameters.filter { it.type != null }.joinToString(",") { it.argumentText }
 
     val signature: String
         get() = buildString {
@@ -150,12 +155,14 @@ class RsExtractFunctionConfig private constructor(
                 .filter { it.textOffset <= end }
 
             val parameters = letBindings
-                .filter { it.textOffset <= start }
-                .filter {
-                    ReferencesSearch.search(it, LocalSearchScope(fn))
-                        .any { ref -> ref.element.textOffset in start..end }
+                .mapNotNull {
+                    if (it.textOffset > start) return@mapNotNull null
+                    val result = ReferencesSearch.search(it, LocalSearchScope(fn))
+                    if (result.none { it.element.textOffset in start..end }) return@mapNotNull null
+                    val requiredBorrowing = result.any { it.element.textOffset > end }
+                        && it.type !is TyPrimitive && !ImplLookup.relativeTo(it).isCopy(it.type)
+                    Parameter.direct(it, requiredBorrowing)
                 }
-                .map { Parameter.direct(it) }
                 .toMutableList()
 
             val innerBindings = letBindings

--- a/src/test/kotlin/org/rust/lang/refactoring/RsExtractFunctionTest.kt
+++ b/src/test/kotlin/org/rust/lang/refactoring/RsExtractFunctionTest.kt
@@ -726,6 +726,62 @@ class RsExtractFunctionTest : RsTestBase() {
         false,
         "bar")
 
+        fun `test extract a function with reference parameter`() = doTest("""
+            #[derive(Copy, Clone, Debug)]
+            struct Copyable;
+
+            fn foo() {
+                let vec = vec![1, 2, 3];
+                let vec2 = vec![1, 2, 3];
+                let mut vec3 = vec![1, 2, 3];
+                let mut vec4 = vec![1, 2, 3];
+                let i = 1;
+                let c = Copyable;
+
+                <selection>println!("{}", vec.len());
+                println!("{}", vec2.len());
+                vec3.push(4);
+                vec4.push(4);
+                println!("{}", i);
+                println!("{:?}", c);</selection>
+
+                println!("{}", vec.len());
+                println!("{}", vec3.len());
+                println!("{}", i);
+                println!("{:?}", c);
+            }
+        """, """
+            #[derive(Copy, Clone, Debug)]
+            struct Copyable;
+
+            fn foo() {
+                let vec = vec![1, 2, 3];
+                let vec2 = vec![1, 2, 3];
+                let mut vec3 = vec![1, 2, 3];
+                let mut vec4 = vec![1, 2, 3];
+                let i = 1;
+                let c = Copyable;
+
+                bar(&vec, vec2, &mut vec3, vec4, i, c);
+
+                println!("{}", vec.len());
+                println!("{}", vec3.len());
+                println!("{}", i);
+                println!("{:?}", c);
+            }
+
+            fn bar(vec: &Vec<i32>, vec2: Vec<i32>, vec3: &mut Vec<i32>, mut vec4: Vec<i32>, i: i32, c: Copyable) {
+                println!("{}", vec.len());
+                println!("{}", vec2.len());
+                vec3.push(4);
+                vec4.push(4);
+                println!("{}", i);
+                println!("{:?}", c);
+            }
+        """,
+        false,
+        "bar")
+
     private fun doTest(@Language("Rust") code: String,
                        @Language("Rust") excepted: String,
                        pub: Boolean,

--- a/src/test/kotlin/org/rust/lang/refactoring/RsExtractFunctionTest.kt
+++ b/src/test/kotlin/org/rust/lang/refactoring/RsExtractFunctionTest.kt
@@ -142,10 +142,10 @@ class RsExtractFunctionTest : RsTestBase() {
         """, """
             fn main() {
                 let mut vec = vec![1, 2, 3];
-                foo(vec);
+                foo(&mut vec);
             }
 
-            fn foo(mut vec: Vec<i32>) {
+            fn foo(vec: &mut Vec<i32>) {
                 vec.push(1);
             }
         """,
@@ -732,23 +732,30 @@ class RsExtractFunctionTest : RsTestBase() {
 
             fn foo() {
                 let vec = vec![1, 2, 3];
-                let vec2 = vec![1, 2, 3];
-                let mut vec3 = vec![1, 2, 3];
-                let mut vec4 = vec![1, 2, 3];
+                let mut mut_vec = vec![1, 2, 3];
                 let i = 1;
-                let c = Copyable;
+                let f = 1.1;
+                let b = true;
+                let c = 'c';
+                let s = "str";
+                let copy = Copyable;
 
                 <selection>println!("{}", vec.len());
-                println!("{}", vec2.len());
-                vec3.push(4);
-                vec4.push(4);
+                mut_vec.push(123);
                 println!("{}", i);
-                println!("{:?}", c);</selection>
+                println!("{}", f);
+                println!("{}", b);
+                println!("{}", c);
+                println!("{}", s);
+                println!("{:?}", copy);</selection>
 
                 println!("{}", vec.len());
-                println!("{}", vec3.len());
                 println!("{}", i);
-                println!("{:?}", c);
+                println!("{}", f);
+                println!("{}", b);
+                println!("{}", c);
+                println!("{}", s);
+                println!("{:?}", copy);
             }
         """, """
             #[derive(Copy, Clone, Debug)]
@@ -756,27 +763,34 @@ class RsExtractFunctionTest : RsTestBase() {
 
             fn foo() {
                 let vec = vec![1, 2, 3];
-                let vec2 = vec![1, 2, 3];
-                let mut vec3 = vec![1, 2, 3];
-                let mut vec4 = vec![1, 2, 3];
+                let mut mut_vec = vec![1, 2, 3];
                 let i = 1;
-                let c = Copyable;
+                let f = 1.1;
+                let b = true;
+                let c = 'c';
+                let s = "str";
+                let copy = Copyable;
 
-                bar(&vec, vec2, &mut vec3, vec4, i, c);
+                bar(&vec, &mut mut_vec, i, f, b, c, s, copy);
 
                 println!("{}", vec.len());
-                println!("{}", vec3.len());
                 println!("{}", i);
-                println!("{:?}", c);
+                println!("{}", f);
+                println!("{}", b);
+                println!("{}", c);
+                println!("{}", s);
+                println!("{:?}", copy);
             }
 
-            fn bar(vec: &Vec<i32>, vec2: Vec<i32>, vec3: &mut Vec<i32>, mut vec4: Vec<i32>, i: i32, c: Copyable) {
+            fn bar(vec: &Vec<i32>, mut_vec: &mut Vec<i32>, i: i32, f: f64, b: bool, c: char, s: &str, copy: Copyable) {
                 println!("{}", vec.len());
-                println!("{}", vec2.len());
-                vec3.push(4);
-                vec4.push(4);
+                mut_vec.push(123);
                 println!("{}", i);
-                println!("{:?}", c);
+                println!("{}", f);
+                println!("{}", b);
+                println!("{}", c);
+                println!("{}", s);
+                println!("{:?}", copy);
             }
         """,
         false,

--- a/src/test/kotlin/org/rust/lang/refactoring/RsExtractFunctionTest.kt
+++ b/src/test/kotlin/org/rust/lang/refactoring/RsExtractFunctionTest.kt
@@ -815,28 +815,36 @@ class RsExtractFunctionTest : RsTestBase() {
         "bar")
 
     fun `test extract a function with passing by &`() = doTest("""
+            fn f(a: &Vec<i32>) {}
+
             fn foo() {
                 let vec = vec![1, 2, 3];
                 let vec2 = vec![1, 2, 3];
+                let vec3 = vec![1, 2, 3];
 
                 <selection>println!("{}", vec.len());
-                println!("{}", vec2.len());</selection>
+                println!("{}", vec2.len());
+                f(&vec3);</selection>
 
                 println!("{}", vec.len());
             }
         """, """
+            fn f(a: &Vec<i32>) {}
+
             fn foo() {
                 let vec = vec![1, 2, 3];
                 let vec2 = vec![1, 2, 3];
+                let vec3 = vec![1, 2, 3];
 
-                bar(&vec, vec2);
+                bar(&vec, vec2, &vec3);
 
                 println!("{}", vec.len());
             }
 
-            fn bar(vec: &Vec<i32>, vec2: Vec<i32>) {
+            fn bar(vec: &Vec<i32>, vec2: Vec<i32>, vec3: &Vec<i32>) {
                 println!("{}", vec.len());
                 println!("{}", vec2.len());
+                f(&vec3);
             }
         """,
         false,
@@ -873,34 +881,28 @@ class RsExtractFunctionTest : RsTestBase() {
     fun `test extract a function with passing by mut`() = doTest("""
             fn test(mut v: Vec<i32>) {}
             fn test2(v: &mut Vec<i32>) {}
-            fn test3(v: Vec<i32>) {}
 
             fn foo() {
                 let mut vec = vec![1, 2, 3];
                 let mut vec2 = vec![1, 2, 3];
-                let vec3 = vec![1, 2, 3];
 
                 <selection>test(vec);
-                test2(&mut vec2);
-                test3(vec3);</selection>
+                test2(&mut vec2);</selection>
             }
         """, """
             fn test(mut v: Vec<i32>) {}
             fn test2(v: &mut Vec<i32>) {}
-            fn test3(v: Vec<i32>) {}
 
             fn foo() {
                 let mut vec = vec![1, 2, 3];
                 let mut vec2 = vec![1, 2, 3];
-                let vec3 = vec![1, 2, 3];
 
-                bar(vec, vec2, vec3);
+                bar(vec, &mut vec2);
             }
 
-            fn bar(mut vec: Vec<i32>, mut vec2: Vec<i32>, vec3: Vec<i32>) {
+            fn bar(mut vec: Vec<i32>, mut vec2: &mut Vec<i32>) {
                 test(vec);
                 test2(&mut vec2);
-                test3(vec3);
             }
         """,
         false,

--- a/src/test/kotlin/org/rust/lang/refactoring/RsExtractFunctionTest.kt
+++ b/src/test/kotlin/org/rust/lang/refactoring/RsExtractFunctionTest.kt
@@ -726,35 +726,75 @@ class RsExtractFunctionTest : RsTestBase() {
         false,
         "bar")
 
-        fun `test extract a function with reference parameter`() = doTest("""
-            #[derive(Copy, Clone, Debug)]
-            struct Copyable;
-
+    fun `test extract a function with passing primitive`() = doTest("""
             fn foo() {
-                let vec = vec![1, 2, 3];
-                let mut mut_vec = vec![1, 2, 3];
                 let i = 1;
                 let f = 1.1;
                 let b = true;
                 let c = 'c';
+
+                <selection>println!("{}", i);
+                println!("{}", f);
+                println!("{}", b);
+                println!("{}", c);</selection>
+
+                println!("{}", i);
+                println!("{}", f);
+                println!("{}", b);
+                println!("{}", c);
+            }
+        """, """
+            fn foo() {
+                let i = 1;
+                let f = 1.1;
+                let b = true;
+                let c = 'c';
+
+                bar(i, f, b, c);
+
+                println!("{}", i);
+                println!("{}", f);
+                println!("{}", b);
+                println!("{}", c);
+            }
+
+            fn bar(i: i32, f: f64, b: bool, c: char) {
+                println!("{}", i);
+                println!("{}", f);
+                println!("{}", b);
+                println!("{}", c);
+            }
+        """,
+        false,
+        "bar")
+
+    fun `test extract a function with passing reference`() = doTest("""
+            fn foo() {
                 let s = "str";
+                <selection>println!("{}", s);</selection>
+                println!("{}", s);
+            }
+        """, """
+            fn foo() {
+                let s = "str";
+                bar(s);
+                println!("{}", s);
+            }
+
+            fn bar(s: &str) {
+                println!("{}", s);
+            }
+        """,
+        false,
+        "bar")
+
+    fun `test extract a function with passing copy trait`() = doTest("""
+            #[derive(Copy, Clone, Debug)]
+            struct Copyable;
+
+            fn foo() {
                 let copy = Copyable;
-
-                <selection>println!("{}", vec.len());
-                mut_vec.push(123);
-                println!("{}", i);
-                println!("{}", f);
-                println!("{}", b);
-                println!("{}", c);
-                println!("{}", s);
-                println!("{:?}", copy);</selection>
-
-                println!("{}", vec.len());
-                println!("{}", i);
-                println!("{}", f);
-                println!("{}", b);
-                println!("{}", c);
-                println!("{}", s);
+                <selection>println!("{:?}", copy);</selection>
                 println!("{:?}", copy);
             }
         """, """
@@ -762,35 +802,105 @@ class RsExtractFunctionTest : RsTestBase() {
             struct Copyable;
 
             fn foo() {
-                let vec = vec![1, 2, 3];
-                let mut mut_vec = vec![1, 2, 3];
-                let i = 1;
-                let f = 1.1;
-                let b = true;
-                let c = 'c';
-                let s = "str";
                 let copy = Copyable;
-
-                bar(&vec, &mut mut_vec, i, f, b, c, s, copy);
-
-                println!("{}", vec.len());
-                println!("{}", i);
-                println!("{}", f);
-                println!("{}", b);
-                println!("{}", c);
-                println!("{}", s);
+                bar(copy);
                 println!("{:?}", copy);
             }
 
-            fn bar(vec: &Vec<i32>, mut_vec: &mut Vec<i32>, i: i32, f: f64, b: bool, c: char, s: &str, copy: Copyable) {
-                println!("{}", vec.len());
-                mut_vec.push(123);
-                println!("{}", i);
-                println!("{}", f);
-                println!("{}", b);
-                println!("{}", c);
-                println!("{}", s);
+            fn bar(copy: Copyable) {
                 println!("{:?}", copy);
+            }
+        """,
+        false,
+        "bar")
+
+    fun `test extract a function with passing by &`() = doTest("""
+            fn foo() {
+                let vec = vec![1, 2, 3];
+                let vec2 = vec![1, 2, 3];
+
+                <selection>println!("{}", vec.len());
+                println!("{}", vec2.len());</selection>
+
+                println!("{}", vec.len());
+            }
+        """, """
+            fn foo() {
+                let vec = vec![1, 2, 3];
+                let vec2 = vec![1, 2, 3];
+
+                bar(&vec, vec2);
+
+                println!("{}", vec.len());
+            }
+
+            fn bar(vec: &Vec<i32>, vec2: Vec<i32>) {
+                println!("{}", vec.len());
+                println!("{}", vec2.len());
+            }
+        """,
+        false,
+        "bar")
+
+    fun `test extract a function with passing by &mut`() = doTest("""
+            fn foo() {
+                let mut vec = vec![1, 2, 3];
+                let mut vec2 = vec![1, 2, 3];
+
+                <selection>vec.push(123);
+                vec2.push(123);</selection>
+
+                println!("{}", vec.len());
+            }
+        """, """
+            fn foo() {
+                let mut vec = vec![1, 2, 3];
+                let mut vec2 = vec![1, 2, 3];
+
+                bar(&mut vec, &mut vec2);
+
+                println!("{}", vec.len());
+            }
+
+            fn bar(vec: &mut Vec<i32>, vec2: &mut Vec<i32>) {
+                vec.push(123);
+                vec2.push(123);
+            }
+        """,
+        false,
+        "bar")
+
+    fun `test extract a function with passing by mut`() = doTest("""
+            fn test(mut v: Vec<i32>) {}
+            fn test2(v: &mut Vec<i32>) {}
+            fn test3(v: Vec<i32>) {}
+
+            fn foo() {
+                let mut vec = vec![1, 2, 3];
+                let mut vec2 = vec![1, 2, 3];
+                let vec3 = vec![1, 2, 3];
+
+                <selection>test(vec);
+                test2(&mut vec2);
+                test3(vec3);</selection>
+            }
+        """, """
+            fn test(mut v: Vec<i32>) {}
+            fn test2(v: &mut Vec<i32>) {}
+            fn test3(v: Vec<i32>) {}
+
+            fn foo() {
+                let mut vec = vec![1, 2, 3];
+                let mut vec2 = vec![1, 2, 3];
+                let vec3 = vec![1, 2, 3];
+
+                bar(vec, vec2, vec3);
+            }
+
+            fn bar(mut vec: Vec<i32>, mut vec2: Vec<i32>, vec3: Vec<i32>) {
+                test(vec);
+                test2(&mut vec2);
+                test3(vec3);
             }
         """,
         false,

